docs/proposals/file-sync-location.md

```diff
@@ -0,0 +1,26 @@
+
+# File Sync Location for Devfile Components in Odo
+
+## Background 
+
+Currently, odo uses the following logic to determine where to sync source code to a devfile component:
+
+1) If the `sourceMapping` field is set, odo will mount the source volume under to the specified folder and sync the source code there
+
+2) If `sourceMapping` is unset, odo will do one of two things:
+   a) If a single project is defined in the devfile, odo will sync to `/projects/<projectName>
+   b) If no projects are defined, or there are multiple projects defined, odo will sync to `/projects`
+
+The downside to this current approach is that the expected sync directory isn't consistent for somebody writing a devfile, and as they add or remove projects from the devfile, the sync directory will change, potentially breaking the commands in their devfile each time.
+
+## Implementation plan
+
+To make the syncing directory more consistent, odo will do the following (in the order given):
+
+1) If the `sourceMapping` field is set, odo will mount the source volume under to the specified folder and sync the source code there. The `PROJECTS_ROOT` environment variable in each container will be set to this value (as-is today).
+
+2) If the `PROJECTS_ROOT` environment variable is set on the user's machine, odo will sync to there, *and* set it as an environment variable in each remore container
+
+3) If neither `sourceMapping` or `PROJECTS_ROOT` are set, odo will mount and sync the project source code to `/projects`
```

Comment:
What if both `sourceMapping` and `PROJECTS_ROOT` are set by the user on purpose or by mistake?

Reply:
Good question. The devfile mentions has the followings (under sourceMapping):
>  sourceMapping: string
    Optional specification of the path in the container where project sources should be transferred/mounted when mountSources is true. When omitted, the value of the PROJECTS_ROOT environment variable is used.

So, the spec doesn't explicitly mention when both `sourceMapping` and `PROJECTS_ROOT` are set, but based off of what it does say, it seems `sourceMapping` would still take priority if both are set?

Reply:
> So, the spec doesn't explicitly mention when both `sourceMapping` and `PROJECTS_ROOT` are set, but based off of what it does say, it seems `sourceMapping` would still take priority if both are set?

+1

I think that description on devfile spec implies that `sourceMapping` takes priority.

Reply:
+1 on sourceMapping has priority